Source/Core/Core/IOS/ES/ES.cpp

```diff
@@ -40,41 +41,38 @@ static u64 s_title_to_launch;
 struct DirectoryToCreate
 {
   const char* path;
-  u32 attributes;
-  OpenMode owner_perm;
-  OpenMode group_perm;
-  OpenMode other_perm;
+  FS::FileAttribute attribute;
+  FS::Mode owner_mode;
+  FS::Mode group_mode;
+  FS::Mode other_mode;
+  FS::Uid uid;
+  FS::Gid gid;
 };
 
 constexpr std::array<DirectoryToCreate, 9> s_directories_to_create = {{
-    {"/sys", 0, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_NONE},
-    {"/ticket", 0, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_NONE},
-    {"/title", 0, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_READ},
-    {"/shared1", 0, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_NONE},
-    {"/shared2", 0, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW},
-    {"/tmp", 0, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW},
-    {"/import", 0, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_NONE},
-    {"/meta", 0, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_RW},
-    {"/wfs", 0, OpenMode::IOS_OPEN_RW, OpenMode::IOS_OPEN_NONE, OpenMode::IOS_OPEN_NONE},
+    {"/sys", 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::None, 0, 0},
+    {"/ticket", 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::None, 0, 0},
+    {"/title", 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::Read, 0, 0},
+    {"/shared1", 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::None, 0, 0},
+    {"/shared2", 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::ReadWrite, 0, 0},
+    {"/tmp", 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::ReadWrite, 0, 0},
+    {"/import", 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::None, 0, 0},
+    {"/meta", 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::ReadWrite, 0x1000, 1},
+    {"/wfs", 0, FS::Mode::ReadWrite, FS::Mode::None, FS::Mode::None, PID_UNKNOWN, PID_UNKNOWN},
 }};
 
 ES::ES(Kernel& ios, const std::string& device_name) : Device(ios, device_name)
 {
   for (const auto& directory : s_directories_to_create)
   {
-    const std::string path = Common::RootUserPath(Common::FROM_SESSION_ROOT) + directory.path;
-
-    // Create the directory if it does not exist.
-    if (File::IsDirectory(path))
-      continue;
-
-    File::CreateFullPath(path);
-    if (File::CreateDir(path))
-      INFO_LOG(IOS_ES, "Created %s (at %s)", directory.path, path.c_str());
-    else
-      ERROR_LOG(IOS_ES, "Failed to create %s (at %s)", directory.path, path.c_str());
-
-    // TODO: Set permissions.
+    const auto result = m_ios.GetFS()->CreateDirectory(0, 0, directory.path, directory.attribute,
```

Comment:
Uid and Gid are currently ignored in the implementation of `CreateDirectory` - should you be passing `directory.uid` and `directory.gid` here to prepare for the time when they'll be used?

Reply:
Here, PID_KERNEL should be used because ES sets its uid/gid to 0/0 so all FS accesses are done as uid 0. I've added a comment to clarify this